react/the_react_ecosystem/fetching-data.md

```diff
@@ -0,0 +1,297 @@
+
+~~~js
+const image = document.querySelector("img");
+fetch("https://jsonplaceholder.typicode.com/photos", {
+  mode: "cors",
+})
+  .then((response) => response.json())
+  .then((response) => {
+    img.src = response[0].url;
+  })
+  .catch((error) => console.error(error));
+~~~
+
+We're making a request to the JSONPlaceholder API to retrieve an image. The response is then converted to JSON format using the json() method, and then we set the image src to the url embedded in that response object. The catch method is used to catch any errors that may occur during the request.
+
+### Using fetch in React components
+
+Now let's take a look at how we can incorporate fetch into a similar React component. One common use case is to fetch data from an API when a component mounts, so that the data can be displayed on screen.
+
+Whenever a component needs to make a request as it renders, it's often best to wrap that fetch inside of an effect.
+
+**https://codesandbox.io/s/top-simple-react-request-rd9psp?file=/src/styles.css**
+
+~~~jsx
+import { useEffect, useState } from "react";
+
+const Image = () => {
+  const [imageURL, setImageURL] = useState(null);
+
+  useEffect(() => {
+    fetch("https://jsonplaceholder.typicode.com/photos", { mode: "cors" })
+      .then((response) => response.json())
+      .then((response) => setImageURL(response[0].url))
+      .catch((error) => console.error(error));
+  }, []);
+
+  return (
+    imageURL && (
+      <>
+        <h1>An image</h1>
+        <img src={imageURL} alt={"placeholder text"} />
+      </>
+    )
+  );
+};
+
+export default Image;
+~~~
+
+useState lets us bundle data within a component and rerender the component whenever state is updated, and useEffect allows us to perform side effects on those components. Since we need to synchronize this component's state (the image URL) with an unpredictable external system like the fetch API, we need an effect. In this case we only want to fetch the URL once when the component is first mounted so we pass an empty array as a dependency to useEffect.
+
+> Reminder: Whenever you are changing state, (or doing anything else that could cause a rerender) inside of useEffect, be sure to always pass a dependency. If you forget to do so useEffect will enter an infinite loop and rerender exponentially.
+
+You might be wondering what's going on here
+
+~~~jsx
+return (
+  imageURL && (
+    <>
+      <h1>An image</h1>
+      <img src={imageURL alt={"placeholder text"}}>
+    </>
+  )
+)
+~~~
+
+This is a pattern called _short circuiting_ and it allows developers to conditionally render JSX without using if statements or ternarys. The above snippet can be read as, "if imageURL is truthy then continue to evaluate the expression". A truthy value and an object evaluates to the object so in this case, JSX is returned. If imageURL was falsy, the right hand side of the && operation wouldn't evaluate and the component would return false.
```

Comment:
All of this the learner has learned in the previous lessons. I say let's remove it.
```suggestion
```